fasm2bels/lib/physical_netlist.py

```diff
@@ -0,0 +1,353 @@
+        self.bel = bel
+        self.pin = pin
+
+        self.branches = []
+
+    def output_interchange(self, obj, string_id):
+        """ Output this route segment and all branches beneth it.
+
+        obj (physical_netlist.RouteBranch pycapnp object) -
+            Object to write PhysicalBelPin into
+        string_id (str -> int) - Function to intern strings into PhysNetlist
+                                 string list.
+
+        """
+        obj.routeSegment.init('belPin')
+        obj.routeSegment.belPin.site = string_id(self.site)
+        obj.routeSegment.belPin.bel = string_id(self.bel)
+        obj.routeSegment.belPin.pin = string_id(self.pin)
+
+        descend_branch(obj, self, string_id)
+
+
+class PhysicalSitePin():
+    """ Python class that represents a site pin in a physical net.
+
+    site (str) - Site containing site pin
+    pin (str) - Site pin in physical net.
+
+    """
+
+    def __init__(self, site, pin):
+        self.site = site
+        self.pin = pin
+
+        self.branches = []
+
+    def output_interchange(self, obj, string_id):
+        """ Output this route segment and all branches beneth it.
+
+        obj (physical_netlist.RouteBranch pycapnp object) -
+            Object to write PhysicalBelPin into
+        string_id (str -> int) - Function to intern strings into PhysNetlist
+                                 string list.
+
+        """
+        obj.routeSegment.init('sitePin')
+        obj.routeSegment.sitePin.site = string_id(self.site)
+        obj.routeSegment.sitePin.pin = string_id(self.pin)
+
+        descend_branch(obj, self, string_id)
+
+
+class PhysicalPip():
+    """ Python class that represents a active pip in a physical net.
+
+    tile (str) - Tile containing pip
+    wire0 (str) - Name of upstream wire to pip
+    wire1 (str) - Name of downstream wire from pip
+    forward (bool) - For bidirectional pips, is the connection from wire0 to
+                     wire1 (forward=True) or wire1 to wire0 (forward=False).
+
+    """
+
+    def __init__(self, tile, wire0, wire1, forward):
+        self.tile = tile
+        self.wire0 = wire0
+        self.wire1 = wire1
+        self.forward = forward
+
+        self.branches = []
+
+    def output_interchange(self, obj, string_id):
+        """ Output this route segment and all branches beneth it.
+
+        obj (physical_netlist.RouteBranch pycapnp object) -
+            Object to write PhysicalBelPin into
+        string_id (str -> int) - Function to intern strings into PhysNetlist
+                                 string list.
+
+        """
+        obj.routeSegment.init('pip')
+        obj.routeSegment.pip.tile = string_id(self.tile)
+        obj.routeSegment.pip.wire0 = string_id(self.wire0)
+        obj.routeSegment.pip.wire1 = string_id(self.wire1)
+        obj.routeSegment.pip.forward = self.forward
+        obj.routeSegment.pip.isFixed = True
+
+        descend_branch(obj, self, string_id)
+
+
+class PhysicalSitePip():
```

Comment:
From my understanding this class is used to model all inverters and routing muxes (like in CLB) in the physical netlist. Am I right?

Reply:
Yes

Reply:
Comments added